File: pypp_cli/test_dir/python/src/file_io/shutil_exceptions.py
import os
import shutil
from pypp_python import pypp_get_resource_dir


def shutil_exceptions_fn():
    print("SHUTIL EXCEPTIONS RESULTS:")
    resources_dir: str = pypp_get_resource_dir()
    try:
        shutil.rmtree(os.path.join(resources_dir, "test_doesn't_exist"))
    except FileNotFoundError as e:
        print("caught FileNotFoundError: ", e)

    try:
        shutil.rmtree(os.path.join(resources_dir, "test_is_here.txt"))
    except NotADirectoryError as e:
        print("caught NotADirectoryError: " + str(e))

    # Be careful with this one that you actually don't have permission to
    # delete the folder.
    # try:
    #     shutil.rmtree(r"C:\Users\puetz\PycharmProjects\bridge-lib-glfw")
    # except PermissionError as e:
    #     print("caught PermissionError: " + str(e))

File: pypp_cli/test_dir/cpp/src/file_io/shutil_exceptions.cpp
#include "file_io/shutil_exceptions.h"
#include "exceptions/filesystem.h"
#include "py_str.h"
#include "pypp_os.h"
#include "pypp_resources.h"
#include "pypp_shutil.h"
#include "pypp_util/print.h"
#include "pypp_util/to_py_str.h"
#include <string>

namespace me {
void shutil_exceptions_fn() {
    pypp::print(pypp::PyStr("SHUTIL EXCEPTIONS RESULTS:"));
    pypp::PyStr resources_dir = pypp::res_dir();
    try {
        pypp::shutil::rmtree(pypp::os::path::join(
            resources_dir, pypp::PyStr("test_doesn't_exist")));
    } catch (const pypp::FileNotFoundError &pypp_pseudo_name_e) {
        std::string e = pypp_pseudo_name_e.msg_;
        pypp::print(pypp::PyStr("caught FileNotFoundError: "), e);
    }
    try {
        pypp::shutil::rmtree(pypp::os::path::join(
            resources_dir, pypp::PyStr("test_is_here.txt")));
    } catch (const pypp::NotADirectoryError &pypp_pseudo_name_e) {
        std::string e = pypp_pseudo_name_e.msg_;
        pypp::print(pypp::PyStr("caught NotADirectoryError: ") + pypp::str(e));
    }
}

} // namespace me